lib/kino/js.ex

```diff
@@ -0,0 +1,362 @@
+                " found in #{inspect(assets_path)}",
+              Macro.Env.stacktrace(env)
+            )
+          end
+
+          loaded_assets
+
+        {true, true} ->
+          IO.warn(
+            "ignoring files in #{inspect(assets_path)} because #{inspect(env.module)}" <>
+              " already defines inline assets with the assets/2 macro",
+            Macro.Env.stacktrace(env)
+          )
+
+          inline_assets
+
+        {false, false} ->
+          message = ~s'''
+          no assets defined for #{inspect(env.module)}.
+
+          Make sure to either explicitly specify assets directory:
+
+              use Kino.JS, assets_path: Path.expand("path/to/assets", __DIR__)
+
+          Or define assets inline:
+
+              asset "main.js" do
+                """
+                export function init(ctx, data) {
+                  ...
+                }
+                """
+              end
+          '''
+
+          IO.warn(message, Macro.Env.stacktrace(env))
+
+          []
+      end
+
+    filenames = Enum.map(assets, &elem(&1, 0))
+    duplicates = Enum.uniq(filenames -- Enum.uniq(filenames))
+
+    if duplicates != [] do
+      duplicates = duplicates |> Enum.map(&inspect/1) |> Enum.join(", ")
+
+      IO.warn(
+        "found duplicate assets in #{inspect(env.module)}: #{duplicates}",
+        Macro.Env.stacktrace(env)
+      )
+    end
+
+    if assets != [] and "main.js" not in filenames do
+      IO.warn(
+        ~s'missing required asset "main.js" in #{inspect(env.module)}',
+        Macro.Env.stacktrace(env)
+      )
+    end
+
+    assets = Enum.uniq_by(assets, &elem(&1, 0))
+
+    dir = dir_for_module(env.module)
+
+    File.rm_rf!(dir)
+    File.mkdir_p!(dir)
+
+    {archive_path, hash} = package_assets!(dir, assets)
+
+    quote do
+      def __js_info__() do
+        %{
+          assets: %{
+            archive_path: unquote(archive_path),
+            js_path: "main.js",
+            hash: unquote(hash)
+          }
+        }
+      end
+    end
+  end
+
+  def __paths__(nil), do: []
+
+  def __paths__(path) do
+    Path.join(path, "**")
+    |> Path.wildcard()
+    |> Enum.reject(&File.dir?/1)
+    |> Enum.map(&String.replace_leading(&1, path <> "/", ""))
+    |> Enum.sort()
+  end
+
+  defp package_assets!(dir_path, assets) do
+    hash = assets_hash(assets)
+    archive_path = Path.join(dir_path, hash <> ".tar.gz")
+
+    archive_content =
+      for {filename, content} <- assets, do: {String.to_charlist(filename), content}
+
+    :ok = :erl_tar.create(archive_path, archive_content, [:compressed])
+
+    {archive_path, hash}
+  end
+
+  defp assets_hash(assets) do
+    md5_hash =
+      assets
+      |> Enum.sort_by(fn {filename, _content} -> filename end)
+      |> Enum.flat_map(&Tuple.to_list/1)
+      |> :erlang.md5()
+
+    Base.encode32(md5_hash, case: :lower, padding: false)
+  end
+
+  defp dir_for_module(module) do
+    priv_dir = :code.priv_dir(:kino)
```

Comment:
Perhaps it is best to ask the user to pass the curernt otp app rather than put everything in Kino? Or is the plan exactly to put it all in Kino to make it easier to read later on?

Reply:
It seems you store an absolute path, so I would definitely do this:

    assets_path: {:current_app, "assets/kino"}

And, after further consideration, I would keep it in lib by default. `assets` directory is not pushed to `hex.pm` by default, so that will bring complication when it comes to publishing. So this:

    assets_path: {:current_app, "lib/mermaid-assets")

Reply:
To be clear, the Kino priv directory is for storing the generated archives, this way we ensure that after compilation all archives are in the Kino priv directory and would be correctly copied to a release, right? Also when defining widgets directly in Livebook, this way we have a known location when the archives would go.

As for the assets path, it currently requires the absolute path yeah, what would `:current_app` resolve to?

Reply:
I see. There is indeed the case with no apps. `:current_app` is no longer relevant if we are always storing in Kino so ignore this issue. Although I would change `assets/mermaid` to `lib/...` something.

Reply:
For Vega-Lite we do `use Kino.JS, assets_path: Path.expand("assets/vega_lite", __DIR__)`, so we target `lib/assets/vega_lite`. Or are you saying we should expand for the user somehow?

Reply:
Oh. I think for clarity you can do “lib/assets/vega_lite”, no need for expand. It will likely make the whole path clearer.

Reply:
How do we get the root URL of the defining package? (and what about in Livebook?)

Reply:
The package is always compiled from its root.

Reply:
Perfect! Just wasn't sure if that's an assumption we can make :D